### PR TITLE
[Decision needed] Added functions to compute DFT and IDFT matrices

### PR DESCRIPTION
**Pull request by oscargus**
#### Brief description of what is fixed or changed
Simple but convenient function to compute the (I)DFT matrix.

#### Other comments
It will be interesting to see if the doctest passes as I have very limited idea how to format the output correctly for this.

#### Release Notes

* discrete
   * Added functions to compute DFT and IDFT matrices

### CODE REVIEW

**Comment by sylee957** [comment]
Dft matrix constructors may better be implemented in `MatrixSpecial` class in `matrices/common.py`, so that it can be possible to construct with different matrix classes easily.

**Comment by oscargus** [comment]
@sylee957 I see your point. I'll have a look.

**Comment by asmeurer** [comment]
CC @sidhantnagpal

**Comment by oscargus** [comment]
I'd be happy to go with any implementation. I have some use cases where it is used on matrix form (to derive fast convolution algorithms), so to me that it is the main thing. It is clearly a bad idea to compute the result of a long DFT this way...

Should I move it to `MatrixSpecial` then? Using `entry`? Or will you @sidhantnagpal add your version?

**Comment by sidhantnagpal** [comment]
Feel free to implement it. The performance can be improved incrementally. 
I should be able to have a look once it is implemented.

> Should I move it to `MatrixSpecial` then? Using `entry`?

I would wait a bit to see more thoughts on this.

**Comment by oscargus** [comment]
I added a version in `MatrixSpecial` now using `entry`. Faster than the previous approach, but probably not as fast as the approach of @sidhantnagpal .

I also changed it to have an optional parameter `inverse` instead of a separate `idft_matrix`method.

Is there a way to reuse the documentation of the two `dft_matrix` methods in `common.py` and `dense.py`?

**Comment by oscargus** [comment]
The documentation in `dense` has a small error in one example. I will push the correct version when it is decided which approach should be kept.

This also indicates that doctest is not running for it as it passed on my computer before pushing. (It also indicates the benefit of having a single source of the documention...)

**Comment by oscargus** [comment]
There are still some things to clean up. I have a local commit, but will not push until later to avoid running too many tests...

**Comment by sylee957** [comment]
I think the implementation on this does not match `fft` in discrete module because of the difference in sign.

However, I think this implementation is more correct than the implementation, and is consistent with numpy, so if it is kept, I think `fft` should be fixed.
I've addressed an issue at #15804

File: sympy/matrices/dense.py [hunk 1]
```diff
@@ -1360,6 +1416,23 @@ def ones(*args, **kwargs):
 
     return Matrix.ones(*args, **kwargs)
 
+def dft_matrix(*args, **kwargs):
+    """Returns a matrix of ones with ``rows`` rows and ``cols`` columns;
+    if ``cols`` is omitted a square matrix will be returned.
```
[the remark this answers is not in the document]
Comment (oscargus) [reply sympy/matrices/dense.py]: This was an old copy of the method that I forgot to remove...

File: sympy/matrices/dense.py [hunk 2]
```diff
@@ -1360,6 +1416,23 @@ def ones(*args, **kwargs):
...
+
+    See Also
+    ========
+
+    zeros
+    eye
+    diag
```
[the remark this answers is not in the document]
Comment (oscargus) [reply sympy/matrices/dense.py]: I'll sort it out in the other methods (which I copied as a template).

File: sympy/matrices/common.py [hunk 3]
```diff
@@ -653,6 +666,73 @@ def entry(i, j):
             return S.Zero
         return cls._new(rows, cols, entry)
 
+    @classmethod
+    def dft_matrix(kls, points, inverse=False, unitary=False, **kwargs):
```
[the remark this answers is not in the document]
Comment (sidhantnagpal) [reply sympy/matrices/common.py]: On second thoughts, I think it maybe fine, as the name of the method name should convey the context.

> 2-D DFT

Most libraries refer to the multidimensional Fourier transform like fft2d/fft2 (or fftn for the n-dimensional case).

**Comment by oscargus** [comment]
Thanks @sidhantnagpal for a thorough review and sorry for the newbie mistakes, but I learned things for the future for sure. I have pushed a new version now.

Regarding the documentation, I simply removed the one in `matrices/common.py`. I tried the `__doc__` approach earlier but didn't get it to work (recursive definitions somehow, do not remember which order I defined it). Seems to be better to have one good source of documentation somehow.

I hope I got everything in it.

**Comment by sylee957** [comment]
In my opinion, the loose functions in `matrices/dense.py` module are the remnants of some old designs, and it may not be a good moment to place some new functions inside.

For example, `ones`, `zeros`, `eye`, `diag`, `jordan_cell`, `multiply_elementwise`, and `GramSchmidt` are just 
 duplicate functions.
And also `carsotian`, `wronskian`, `hessian` may have to be placed in the similar context as `Matrix.jacobian`.

Once the special matrix constructors are set as classmethods, it is possible to use them without creating an instance, and users can take abbreviations like `dft = Matrix.dft_matrix` or whatever if they need, without much effort.

**Comment by oscargus** [comment]
I just found sympy.matrices.expressions.fourier which contains DFT and IDFT. Hence, this is not required. There is still a bit of potential to improve those, so I spend my efforts there.

**Pull request closed without merging**